csrc/host_ir/jit.cpp

```diff
@@ -0,0 +1,383 @@
+          communicator,
+          hir::HostIrEvaluatorParams{},
+          num_threads) {}
+
+HostIrJit::HostIrJit(
+    hir::HostIrContainer* container,
+    Communicator* communicator,
+    const hir::HostIrEvaluatorParams& params,
+    int num_threads)
+    : pimpl_(new LlvmJitImpl) {
+  // Initialize params with passed parameters
+  pimpl_->params_ =
+      std::make_unique<HostIrJitParams>(container, communicator, params);
+
+  llvm::InitializeNativeTarget();
+  llvm::InitializeNativeTargetAsmPrinter();
+  pimpl_->jit = ExitOnErr(
+      llvm::orc::LLJITBuilder().setNumCompileThreads(num_threads).create());
+  llvm::orc::JITDylib& dest_dynamic_lib = pimpl_->jit->getMainJITDylib();
+  auto mangler = llvm::orc::MangleAndInterner(
+      dest_dynamic_lib.getExecutionSession(), pimpl_->jit->getDataLayout());
+  dest_dynamic_lib.addGenerator(
+      ExitOnErr(llvm::orc::DynamicLibrarySearchGenerator::GetForCurrentProcess(
+          pimpl_->jit->getDataLayout().getGlobalPrefix())));
+
+  // Create wrapper function pointer to at::native::empty_strided_cuda
+  void* empty_strided_cuda_func_ptr = reinterpret_cast<void*>(
+      +[](const int64_t* sizes,
+          int64_t ndim,
+          const int64_t* strides,
+          int64_t strides_ndim,
+          int32_t dtype,
+          int64_t device_index) -> at::Tensor* {
+        at::IntArrayRef aten_sizes(sizes, ndim);
+        at::IntArrayRef aten_strides(strides, strides_ndim);
+        // Use the type and device passed as parameters
+        at::ScalarType scalar_type = static_cast<at::ScalarType>(dtype);
+        at::Device device =
+            at::Device(at::kCUDA, static_cast<c10::DeviceIndex>(device_index));
+        return new at::Tensor(at::native::empty_strided_cuda(
+            aten_sizes,
+            aten_strides,
+            scalar_type,
+            c10::nullopt,
+            device,
+            c10::nullopt));
+      });
+
+  // Register at::native::empty_strided_cuda function in LLVM
+  auto empty_strided_cuda_addr =
+      llvm::orc::ExecutorAddr::fromPtr(empty_strided_cuda_func_ptr);
+  llvm::orc::SymbolMap symbolMap;
+  symbolMap[mangler("at::native::empty_strided_cuda")] = llvm::orc::ExecutorSymbolDef(
+      empty_strided_cuda_addr, llvm::JITSymbolFlags::Exported);
+
+  ExitOnErr(dest_dynamic_lib.define(llvm::orc::absoluteSymbols(symbolMap)));
+  
+  // Only compile if container is provided
+  if (container) {
+    compile(container, pimpl_->jit.get(), pimpl_->allocate_funcs_, *pimpl_->params_);
+  }
+}
+
+HostIrJit::~HostIrJit() = default;
+
+at::Tensor HostIrJit::allocate(
+    const kir::Allocate* allocate,
+    const std::vector<int64_t>& input_sizes,
+    const std::vector<int64_t>& input_strides) {
+  if (pimpl_->allocate_funcs_.find(allocate) == pimpl_->allocate_funcs_.end()) {
+    NVF_ERROR(false, "allocate function not found for ", allocate);
+  }
+  FUSER_PERF_SCOPE("HostIrJit::allocate");
+  auto func_ptr = pimpl_->allocate_funcs_[allocate];
```

Comment:
nitpick: let's avoid double cache lookup.
```suggestion
  auto allocate_funcs_iter = pimpl_->allocate_funcs_.find(allocate);
  NVF_ERROR(allocate_funcs_iter != pimpl_->allocate_funcs_.end(), "allocate function not found for ", allocate);
```

I think you can replace `func_ptr` with `allocate_funcs_iter->second` below.